ipcache: Fix handling of endpoint IP events

rlenglet opened the pull request:
Move the slice of ipcache listeners from `Daemon` into `IPCache`.
Factorize and encapsulate the handling of listeners into `IPCache`'s `Upsert` and `Delete` methods to reduce the complexity of callers.

Modify the signature of the `OnIPIdentityCacheChange` method in the `IPIdentityMappingListener` interface.
Specify that the given IP is always in CIDR form, with a prefix length, even when it represents an endpoint IP.
Use stronger types in the signature to clarify that only the identity and host IP associated with the CIDR may be modified.

Fix the `Upsert` and `Delete` to handle all cases of CIDR shadowing by endpoint IP:

- If inserting a new CIDR that is shadowed by an existing endpoint IP, don't callback the listeners about the insert.
- If inserting a new endpoint IP that is shadowing an existing CIDR, notify the listeners of an update of the identity associated with the CIDR, instead of an insert. If the shadowed CIDR and the endpoint IP are associated with the same identity, don't callback.
- If deleting a CIDR that was shadowed by an endpoint IP, don't callback the listeners about the delete.
- If deleting an endpoint IP that was shadowing a CIDR, notify the listeners of an update of the identity associated with the CIDR, instead of a delete. If the shadowed CIDR and the endpoint IP were associated with the same identity, don't callback.

In `Upsert`, when updating an IP's identity, delete the old identity from the identity-to-IP map.

Cache the mappings from endpoint IP / CIDR to host IP (maybe `nil`), in order to detect when the host IP for an endpoint changes.
Don't bypass the ipcache listener notifications when the host IP changes but the identity is identical.

Fixes: https://github.com/cilium/cilium/issues/4803
Fixes: https://github.com/cilium/cilium/issues/4798
Fixes: #3976

Signed-off-by: Romain Lenglet <romain@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4810)
<!-- Reviewable:end -->

rlenglet committed:
ipcache: Fix handling of endpoint IP events

Move the slice of ipcache listeners from Daemon into IPCache.
Factorize and encapsulate the handling of listeners into
IPCache's Upsert and Delete methods to reduce the complexity
of callers.

Modify the signature of the OnIPIdentityCacheChange method in the
IPIdentityMappingListener interface.
Specify that the given IP is always in CIDR form, with a prefix
length, even when it represents an endpoint IP.
Use stronger types in the signature to clarify that only the
identity and host IP associated with the CIDR may be modified.

Fix the Upsert and Delete to handle all cases of CIDR shadowing
by endpoint IP.
If inserting a new CIDR that is shadowed by an existing endpoint
IP, don't callback the listeners about the insert.
If inserting a new endpoint IP that is shadowing an existing
CIDR, notify the listeners of an update of the identity associated
with the CIDR, instead of an insert. If the shadowed CIDR and the
endpoint IP are associated with the same identity, don't callback.
If deleting a CIDR that was shadowed by an endpoint IP, don't
callback the listeners about the delete.
If deleting an endpoint IP that was shadowing a CIDR, notify the
listeners of an update of the identity associated with the CIDR,
instead of a delete. If the shadowed CIDR and the endpoint IP were
associated with the same identity, don't callback.

In Upsert, when updating an IP's identity, delete the old identity
from the identity-to-IP map.

Fixes: https://github.com/cilium/cilium/issues/4803
Fixes: https://github.com/cilium/cilium/issues/4798
Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+		ones, bits := cidr.Mask.Size()
+		if ones == bits {
+			if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+				scopedLog.Info("Upserting CIDR shadowed by endpoint IP")
```
Let's make this info message more readable for an end user:
`Ignoring CIDR to identity mapping as it would shadow an endpoint IP`

rlenglet replied:
Done.

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+		if !found {
+			if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+				if cidrIdentity.ID != newIdentity.ID {
+					scopedLog.Info("Upserting endpoint IP shadowing CIDR")
```
`New endpoint IP started shadowing existing CIDR to identity mapping`

rlenglet replied:
Done.

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+				} else {
+					// The endpoint IP and the CIDR are associated with the
+					// same identity. Nothing changes for the listeners.
+					callbackListeners = false
```
This seems wrong as the endpoint IP based mapping can have a different hostIP that needs to be plumbed.

rlenglet replied:
This is not worse than the current behavior. Currently, if we call `Upsert` with the same IP and same ID, but a different host IP, the listeners won't be called back either.

My take is that this is OK since the host IP for an endpoint IP will never change, given how we're allocating endpoint IPs in Cilium. And the initial association between an endpoint IP and its host IP will always be forwarded to the listeners. So it will continue to work as it does now.

If we wanted ipcache to be responsible for mapping endpoint IPs to host IPs, the logic to determine whether to callback the listeners and with what parameters should have been different and more complicated. Notably, we should maintain a map of IP to host IP, because we'd need to keep track of the old host IP so it could be passed to the listeners and so we could determine whether to callback the listeners when the host IP changes.

rlenglet replied:
Just to clarify: in which cases does the host IP for an endpoint IP change? I thought that this couldn't happen in Cilium.

joestringer replied:
Is it a bad thing to complain loudly in the logs here if the hostIP doesn't match?

tgraf replied:
I'm happy with a loud log message for now but would feel more comfortable if we hand it gracefully at some point. I'm sure we'll forget about the assumption we're making here.

rlenglet replied:
Unfortunately, to be able to detect a mismatch I’ll have to maintain a map of IP to host IP. If we’re biting the bullet and manage that mapping anyway, we might as well do it the right way.

ianvernon replied:
I don't understand the concern that " the endpoint IP based mapping can have a different hostIP that needs to be plumbed" - when can this be the case? When could such a mismatch occur?

tgraf replied:
The most likely scenario are outdated ipcache entries in the kvstore that then get overwritten by a node reusing a previously used PodCIDR with a changed hostIP. In general I find it highly dangerous to make assumptions in such generic code that is at the foundation of everything. We can exclude this case for now and track it in an issue but we should really turn it around and proof that a hostIP can never change instead of arguing about likeliness of it changing.

rlenglet replied:
I updated ipcache to properly handle changing the host IP for an endpoint.

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+		// this case, skip calling back the listeners since they don't know
+		// about its mapping.
+		if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+			scopedLog.Info("Deleting CIDR shadowed by endpoint IP")
```
Why does this need a info message?

rlenglet replied:
For some reason, the shadowing-related log messages were at info level. I just wanted to be consistent. I agree this should be debug.
Changed all of them.

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+		// restore its mapping with the listeners if that was the case.
+		if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+			if cidrIdentity.ID != cachedIdentity.ID {
+				scopedLog.Info("Deleting endpoint IP shadowing CIDR, restoring CIDR")
```
Removal of endpoint IP revives shadowed CIDR to identity mapping

tgraf commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+			} else {
+				// The endpoint IP and the CIDR were associated with the same
+				// identity. Nothing changes for the listeners.
+				callbackListeners = false
```
This needs to take into account a potential change in the hostIP as well. Otherwise we may continue to encap.

rlenglet replied:
Yes, agreed. :/
I think this will require keeping a mapping of IP to host IP, so we know whether the host IP changes or not because of un/shadowing.
The daemon callback will need to be changed as well, since this callback wouldn't be a `Delete` but an `Upsert` with a `nil` new host IP.

joestringer replied:
Should hostIP changes be a separate cache modification type? Then the proxy can ignore it easily. Only the BPF side needs to be aware of such changes, right?

rlenglet replied:
I’ve been thinking about this. I would like that very much. However I don’t think that would be possible since the ID and host IP need to change together to be able to compute the correct BPF map key, cf the callback implementation in the daemon.

tgraf reviewed (changes requested):
Great change and I understand the fix of detecting shadowed CIDRs.

I don't understand yet how this fixes #4808 as the listener update is only skipped if the identity is already equal. In the commit message it says:

```
If inserting a new endpoint IP that is shadowing an existing
CIDR, notify the listeners of an update of the identity associated
with the CIDR, instead of an insert.
```

Where is the code that does that?

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+			}
+		}
+	} else {
+		scopedLog.Debug("Attempt to upsert invalid IP into ipcache layer")
```
`.WithError(err)`?

rlenglet replied:
I have no `error` to work with. When `net.ParseIP` can't parse, it just returns a `nil` value, and no error.
I made this an `Error` log statement.

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+	callbackListeners := true
 
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
```
This logic (the next ~40-50LoC) feels like it should be able to be shared with `Upsert()`.

joestringer commented on pkg/ipcache/kvstore.go:
```diff
@@ -362,106 +338,37 @@ restart:
+					scopedLog.WithError(err).Errorf("Not adding entry to ip cache; error unmarshaling data from key-value store")
 					continue
 				}
+				IPIdentityCache.Upsert(ipIDPair.PrefixString(), ipIDPair.HostIP, Identity{
```
:tada:

joestringer commented on pkg/ipcache/kvstore.go:
```diff
@@ -362,106 +338,37 @@ restart:
 			switch event.Typ {
 			case kvstore.EventTypeListDone:
-				for _, listener := range listeners {
+				for _, listener := range IPIdentityCache.listeners {
```
Unlocked access to the listeners?

EDIT: Specifically, I believe that `InitIPIdentityWatcher()`->`Watch()` here  is called prior to `SetListeners`, so this could be running while the daemon sets the listeners.

rlenglet replied:
Added locking around this. Good catch!

`SetListeners` is called from `Daemon.init`, which is called prior to `ipcache.InitIPIdentityWatcher`, cf. `NewDaemon`:
```
	if err = d.init(); err != nil {
		log.WithError(err).Error("Error while initializing daemon")
		return nil, err
	}

	// Start watcher for endpoint IP --> identity mappings in key-value store.
	// this needs to be done *after* init() for the daemon in that function,
	// we populate the IPCache with the host's IP(s).
	ipcache.InitIPIdentityWatcher()
```

joestringer reviewed (commented):
+5 LoC and so much tidier. Glad to see this!

I added some minor feedback, in particular one shared variable access without holding a lock.

rlenglet committed:
ipcache: Maintain IP to host IP cache

Cache the mappings from endpoint IP / CIDR to host IP (maybe nil),
in order to detect when the host IP for an endpoint changes.

Don't bypass the listener notifications when the host IP changes
but the identity is identical.

Signed-off-by: Romain Lenglet <romain@covalent.io>
